R/expectations-matches.R

```diff
@@ -126,39 +129,58 @@ prints_text <- function(regexp, ...) {
 
 #' @export
 #' @rdname matching-expectations
-expect_error <- function(object, regexp = NULL, ..., info = NULL,
-  label = NULL) {
+expect_error <- function(object, regexp = NULL, expected_class = NULL, ..., info = NULL,
+                         label = NULL) {
   if (is.null(label)) {
     label <- find_expr("object")
   }
-  expect_that(object, throws_error(regexp, ...), info = info, label = label)
+  expect_that(object, 
+              throws_error(regexp, expected_class = expected_class, ...),
+              info = info, label = label)
 }
 #' @export
 #' @rdname oldskool
-throws_error <- function(regexp = NULL, ...) {
+throws_error <- function(regexp = NULL, expected_class = NULL, ...) {
+  if (identical(regexp, NA) && !is.null(expected_class)) {
+    stop("If regexp argument is NA (expecting no error), class should be NULL")
+  }
   function(expr) {
-    res <- try(force(expr), TRUE)
-
-    if (inherits(res, "try-error")) {
-      errors <- as.character(res)
-    } else {
-      errors <- character()
-    }
-
+    error <- tryCatch({
+      expr
+      NULL
+    },
+    error = function(e) {
+      e
+    })
+    
+    paste_with_commas <- function(x, y) paste(x, y, sep=", ")
+    expected_classes_as_string <- Reduce(paste_with_commas, expected_class)
+    no_error_string <- "no errors raised"
+    
     if (identical(regexp, NA)) {
       expectation(
-        length(errors) == 0,
-        paste0("expected no errors:\n", paste("* ", errors, collapse = "\n")),
-        "no errors raised"
+        is.null(error),
+        paste0("expected no errors:\n", as.character(error),
+               no_error_string),
+        success_msg = no_error_string
       )
-    } else if (!is.null(regexp) && length(errors) > 0) {
-      matches(regexp, ...)(errors)
-    } else {
-      expectation(
-        length(errors) > 0,
-        "no errors raised",
-        paste0(length(errors), " errors raised")
+    } else if (is.null(error)) {
+      expectation(FALSE, no_error_string) 
```

Comment:
You're right it's confusing - I did it the way I did b/c I thought the failure message should depend on whether there's no error, or an error of the wrong type. So some branching on `!is.null(error)` is required (unless we want to give up on distinguishing in the message), but I can make it cleaner.